eth/connmanager.go

```diff
@@ -0,0 +1,254 @@
+			return !p.DynDialed() ||
+				p.Trusted() ||
+				p.Lifetime() < mclock.AbsTime(doNotDropBefore)
+		} else {
+			// Only drop from inbound peers. Avoid dropping trusted peers.
+			// Give some time to peers before considering them for a drop.
+			return p.DynDialed() || p.StaticDialed() ||
+				p.Trusted() ||
+				p.Lifetime() < mclock.AbsTime(doNotDropBefore)
+		}
+	}
+	droppable := slices.DeleteFunc(peers, selectDoNotDrop)
+	if len(droppable) > 0 {
+		p := droppable[cm.rand.Intn(len(droppable))]
+		cm.log.Debug("dropping random peer", "id", p.ID(), "duration", common.PrettyDuration(p.Lifetime()),
+			"dialed", dialed, "peercountbefore", len(peers))
+		p.Disconnect(p2p.DiscTooManyPeers)
+		return true
+	}
+	return false
+}
+
+// randomDuration generates a random duration between min and max.
+// TODO: maybe we should move this to a common utlity package.
+// TODO: panic might be too harsh, maybe return an error.
+func randomDuration(rand *mrand.Rand, min, max time.Duration) time.Duration {
+	if min > max {
+		panic("min duration must be less than or equal to max duration")
+	}
+	nanos := rand.Int63n(max.Nanoseconds()-min.Nanoseconds()) + min.Nanoseconds()
+	return time.Duration(nanos)
+}
+
+// updatePeerDropTimers checks and starts/stops the timer for peer drop.
+func (cm *connManager) updatePeerDropTimers(syncing bool) {
+	numPeers, numDialed, numInbound := cm.numPeers()
+	cm.log.Trace("ConnManager status", "syncing", syncing,
+		"peers", numPeers, "out", numDialed, "in", numInbound,
+		"maxout", cm.maxDialPeers, "maxin", cm.maxInboundPeers)
+
+	if !syncing {
+		// If a drop was already scheduled, Schedule does nothing.
+		if cm.maxDialPeers-numDialed <= peerDropThreshold {
+			interval := randomDuration(cm.rand, peerDropIntervalMin, peerDropIntervalMax)
+			cm.peerDropDialedTimer.Schedule(cm.clock.Now().Add(interval))
+		} else {
+			cm.peerDropDialedTimer.Stop()
+		}
+
+		if cm.maxInboundPeers-numInbound <= peerDropThreshold {
+			interval := randomDuration(cm.rand, peerDropIntervalMin, peerDropIntervalMax)
+			cm.peerDropInboundTimer.Schedule(cm.clock.Now().Add(interval))
+		} else {
+			cm.peerDropInboundTimer.Stop()
+		}
+	} else {
+		// Downloader is managing connections while syncing.
+		cm.peerDropDialedTimer.Stop()
+		cm.peerDropInboundTimer.Stop()
+	}
+}
+
+// loop is the main loop of the connection manager.
+func (cm *connManager) loop() {
+	defer cm.wg.Done()
+
+	// Set up periodic timer to pull syncing status.
+	// We could get syncing status in a few ways:
+	// - poll the sync status (we use this for now)
+	// - subscribe to Downloader.mux
+	// - subscribe to DownloaderAPI (which itself polls the sync status)
+	syncing := cm.syncingFunc()
+	cm.log.Trace("Sync status", "syncing", syncing)
+	syncCheckTimer := mclock.NewAlarm(cm.connmanConfig.clock)
+	syncCheckTimer.Schedule(cm.clock.Now().Add(syncCheckInterval))
+	defer syncCheckTimer.Stop()
+
+	for {
+		select {
+		case <-syncCheckTimer.C():
```

Comment:
Process sync events here:

```
case ev <- cm.syncsub.Chan():
     switch ev := ev.(type) {
     case downloader.StartEvent:
         ...
     case downloader.StopEvent:
         ...
     }
```

Reply:
With the new version we can just read the current sync status as part of the decision logic.